magda-gateway/src/Authenticator.ts

```diff
@@ -43,10 +109,146 @@ export default class Authenticator {
         this.passportSessionMiddleware = passport.session();
     }
 
+    /**
+     * destroy the session. Will do two things:
+     * - Delete session data from session store
+     * - Delete session cookie
+     *
+     * @param {express.Request} req
+     * @param {express.Response} res
+     * @returns {Promise<never>}
+     * @memberof Authenticator
+     */
+    destroySession(
+        req: express.Request,
+        res: express.Response
+    ): Promise<never> {
+        return new Promise((resolve, reject) => {
+            req.session.destroy(err => {
+                if (err) {
+                    // Failed to access session storage to delete session data
+                    reject(err);
+                } else {
+                    resolve();
+                }
+            });
+            const deleteCookieOptions = {
+                ...this.sessionCookieOptions
+            };
+            // --- `clearCookie` works in a way like it will fail to delete cookie if maxAge presents T_T
+            // --- https://github.com/expressjs/express/issues/3856#issuecomment-502397226
+            delete deleteCookieOptions.maxAge;
+            res.clearCookie(DEFAULT_SESSION_COOKIE_NAME, deleteCookieOptions);
+        });
+    }
+
+    /**
+     * A middleware wraps all other cookie / session / passport related middlewares
+     * to achieve fine-gain session / cookie control in Magda.
+     * Generally, we want to:
+     * - user request won't carry a cookie in the header unless they are logged in
```

Comment:
I was trying to say we want to make sure users' requests from the web browser will not carry `Cookie` header.
At this moment, even the user is logout, it will still carry it.
I probably missed your point. 
Can you specify a bit more details?

Reply:
Ah right. I was a bit confused because we don't actually have any control over the request, just the response. 

So what you're trying to say is that we want to make sure that the user's browser doesn't get a cookie set on it until they log in, and that the cookie gets un-set when they log out?

Reply:
Yes~ So there are two parts:
- The client does not carry cookie (i.e. their request will not include Cookie header)
- The server stops always sending `setCookie` header (at this moment, it always do)